fix(ivy): ensure that all NGCC switch markers are processed correctly

petebacondarwin opened the pull request:
This PR is a big refactoring - related to the discussion at https://github.com/angular/angular/pull/26070#issuecomment-423960801.

It sits on top of #26236.

petebacondarwin committed:
refactor(ivy): move ngcc `rootDirs` computation into a function

mary-poppins commented:
You can preview a833630 at https://pr26082-a833630.ngbuilds.io/.

mary-poppins commented:
You can preview 99592a2 at https://pr26082-99592a2.ngbuilds.io/.

JoostK commented:
This makes a lot of sense! Although you mention no behavior changes, looks like this also fixes #26070 I was trying to fix.

petebacondarwin commented:
LOL sorry that was not my intention! I certainly havent tested that :-)
Would you like to use it as a basis for a fully tested fix?

petebacondarwin commented:
Moreover I would welcome your feedback on the approach. I still feel awkward that DTS stuff is so of its own thing while the renderer knows about classes and markers.

JoostK commented on packages/compiler-cli/src/ngcc/src/analysis/switch_marker_analyzer.ts:
```diff
@@ -0,0 +1,34 @@
+import * as ts from 'typescript';
+import {NgccReflectionHost, SwitchableVariableDeclaration} from '../host/ngcc_host';
+
+export interface SwitchMarkerAnalysis {
```
Nice choice for the name, I was struggling to come up with something descriptive and succinct :-)

JoostK commented on packages/compiler-cli/src/ngcc/src/analysis/switch_marker_analyzer.ts:
```diff
@@ -0,0 +1,34 @@
+   */
+  analyzeEntryPoint(program: ts.Program): Map<ts.SourceFile, SwitchMarkerAnalysis> {
+    const analyzedFiles = new Map<ts.SourceFile, SwitchMarkerAnalysis>();
+    program.getSourceFiles().forEach(sourceFile => {
```
I think this is where my bug is fixed, since the approach of iterating over all source files is the main idea of the fix.

JoostK commented on packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -60,11 +60,25 @@ export class Analyzer {
+   * @param entryPoint The file where the analysis should begin.
+   * @return A map of TS source files to analyzed file info.
+   */
+  analyzeEntryPoint(entryPoint: ts.SourceFile): Map<ts.SourceFile, DecorationAnalysis> {
```
Seems strange to me that this takes a sourcefile, whereas switch marker analyzer takes the whole program. That means the caller now defines the analysis scope of the analyzer. Given that they conceptually fall under the umbrella of analyzers as you proposed, they should adhere to the same interface.

We don't necessarily need an interface now, though (doesn't really add anything, only complicates matters).

petebacondarwin replied:
Ahah! You spotted my deliberate mistake :-)
Yes I think it would be better to pass in the `program`, which I think allows us to get the same `SourceFlle` via `program.getRootFileNames()[0]`.

JoostK commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -117,6 +123,42 @@ export class Esm5ReflectionHost extends Fesm2015ReflectionHost {
+   * @returns A collection of files objects that hold info about the decorated classes and import
+   * information.
+   */
+  findDecoratedFiles(entryPoint: ts.SourceFile): Map<ts.SourceFile, DecoratedFile> {
```
Moving these to the reflection hosts makes a whole lot of sense!

JoostK commented on packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -60,11 +60,25 @@ export class Analyzer {
+   */
+  analyzeEntryPoint(entryPoint: ts.SourceFile): Map<ts.SourceFile, DecorationAnalysis> {
+    const analyzedFiles = new Map<ts.SourceFile, DecorationAnalysis>();
+    const decoratedFiles = this.host.findDecoratedFiles(entryPoint);
```
I was mainly struggling with how the result of `FileParser` was used to call an analyzer multiple times, so inverting the dependency here makes it much, much simpler to let separate analyzers are responsible for their own analysis scope.

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -65,14 +63,7 @@ export class Transformer {
-    } else {
-      rootDirs = [host.getCurrentDirectory()];
-    }
+    const rootDirs = this.getRootDirs(host, options);
```
💯

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -82,27 +73,59 @@ export class Transformer {
-    const analyzer = new Analyzer(typeChecker, reflectionHost, rootDirs);
+    // Analyze the files for NGCC switch marker
+    const switchMarkerAnalyzer = new SwitchMarkerAnalyzer(reflectionHost);
+    const switchMarkerAnalyses = switchMarkerAnalyzer.analyzeEntryPoint(packageProgram);
```
Calling the analyzers separately here makes it a bit easier, but makes transformer quite convoluted. I opted for a single "main" analyzer that delegated to the decorator analyzer  such that given main analyzer could merge the results together, resulting in a single list of analysis results.

I think it would still makes sense to introduce a layer that collects the results from all analyzers and merges them into a single list of results, similar to what I did before.

petebacondarwin replied:
I think you are right. I put it all in the Transformer because, in my mind, that is the only place that ought to know about all the different aspects of transformation.

So, the problem I am fighting with, which I would really like to avoid, is to have lots of classes that each know about all the different parsing/analysis/rendering requirements.

So at the moment Hosts are nicely unaware of the analysis and rendering tasks.

But let's say we go with a general Analyzer class, then that class would need to know about marker analysis AND decorator analysis (and potentially .d.ts analysis if we went the full journey). And the Renderer would also need to know about marker rendering and decorator rendering (and potentially .d.ts rendering if we went the full journey).

This results in a number of classes that all have lots of knowledge of large parts of the system.

Ideally I would like some way to limit the knowledge of such classes. This is what I was talking about before in terms of a plug-in approach.

JoostK replied:
Ah, I see, yes I agree.

How important is is for analyzers to have access to multiple source files? If not at all, we may opt for a design similar to TypeScript's transformers: have an umbrella "processor" that calls its analyzer, then uses that result to call its own renderer. The transformer could simply iterate all processors in order for all source files, all of them operating independently, after which it writes the transformed source file to disk.

Well, you wrote dgeni, which this basically boils down to I guess 😛

petebacondarwin replied:
Analyzers do need to access the whole `program` because they may need to resolve computed values that span multiple source files.

JoostK replied:
Sure, that shouldn't be a problem I think. The only requirement I see is that the processing of source files must happen independently from each other (analysis of one file should not have to be used by another analyzer)

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -125,19 +158,6 @@ export class Transformer {
     }
   }
 
-  getFileParser(format: string, program: ts.Program, host: NgccReflectionHost): FileParser {
```
🎉

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -82,27 +73,59 @@ export class Transformer {
+      // Transform the `.d.ts` files.
+      // TODO(gkalpak): What about `.d.ts` source maps? (See
+      // https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-9.html#new---declarationmap.)
+      if (decoratorAnalysis && format === 'esm2015') {
```
I like how each source file is now processed as a single unit 👍

JoostK commented on packages/compiler-cli/src/ngcc/test/rendering/esm2015_renderer_spec.ts:
```diff
@@ -8,23 +8,28 @@
+    decoratorAnalyzer: DecorationAnalyzer, switchMarkerAnalyzer: SwitchMarkerAnalyzer,
+    program: ts.Program) {
+  const file = program.getSourceFile(PROGRAM.name) !;
+  return {
```
Much better than before 😃

JoostK reviewed (commented):
These are much welcomed improvements! Because of the changes to how the analysis scope of switch markers is now determined, looks like this already fixes #26070.

I'm not sure I understand you concern regarding the DTS files and the renderer.

petebacondarwin commented:
@JoostK now that really is a thorough review. Thank you!

I would still like to work out how to avoid my concerns in https://github.com/angular/angular/pull/26082#discussion_r219947405

But perhaps we should consider getting this in, with your suggested fixes and some more unit tests for the broken usage case that prompted your original PR.

JoostK commented:
@petebacondarwin I would propose to get the bug fixed separate from more rework, even more so as the desired design hasn't fully materialized yet.

I'm quite busy for the remaining days this week, so I won't have much time to work on this until the weekend.

petebacondarwin committed:
refactor(ivy): implement `NgccReflectionHost.findDecoratedFiles`

petebacondarwin committed:
refactor(ivy): rename and move ngcc `Parsed...` to `Decorated...`

petebacondarwin committed:
refactor(ivy): move and rename `Analyzer` to `DecorationAnalyzer`

This is in preparation for adding in other kinds of Analyzer.

petebacondarwin committed:
refactor(ivy): ngcc - `DecorationAnalyzer` acts on whole program

petebacondarwin committed:
refactor(ivy): ngcc - `Transformer` delegates to `Analyzer`s and `Renderer`

petebacondarwin commented:
FYI I have rebased on top of #25905 - but will need another rebase soon as I am about to update that.
It was quite a painful rebase, as I had to rebuild each commit from scratch, but it gave me an opportunity to think a bit further about the approach.

mary-poppins commented:
You can preview 9f17f65 at https://pr26082-9f17f65.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/rendering/fesm2015_renderer.ts:
```diff
@@ -0,0 +1,105 @@
+  constructor(
+      protected program: ts.Program, protected host: NgccReflectionHost,
+      protected decorationAnalyses: DecorationAnalyses,
+      protected switchMarkerAnalyses: SwitchMarkerAnalyses, protected sourcePath: string,
```
It turns out that the only thing that the `Renderer` does is `renderProgram`. And it only does that exactly once in its lifetime. So really all the stuff I am passing in to the constructor are only needed for the `renderProgram` call, so they could have equally be put there too. But then we would have had to either attach them to the class at some point or pass them around between the methods. It seemed weird to be in a half-way house where we passed some things into the constructor and others in to the call to `renderProgram` so I went to one extreme.
Perhaps the other extreme is better - to make the `Renderer` completely stateless, and so pass everything around between the methods.
What do you think?

petebacondarwin replied:
But now that the `Esm2015Renderer` overrides the `renderFile` method and requires a `DtsMapper` object, to be provided (via its constructor) then there will always need to be something on the constructor.

I guess a compromise of the host, sourcePath, targetPath and dtsMapper on the constructor is acceptable...

petebacondarwin replied:
Updated

mary-poppins commented:
You can preview 0c8f993 at https://pr26082-0c8f993.ngbuilds.io/.

petebacondarwin committed:
refactor(ivy): remove ngcc `Parser` and use `NgccReflectionHost` instead

mary-poppins commented:
You can preview 84a2c01 at https://pr26082-84a2c01.ngbuilds.io/.

petebacondarwin committed:
refactor(ivy): implement ngcc `SwitchMarkerAnalyzer`

mary-poppins commented:
You can preview 220b799 at https://pr26082-220b799.ngbuilds.io/.

petebacondarwin committed:
refactor(ivy): ngcc - `Renderer` now manages d.ts transformation

mary-poppins commented:
You can preview 1c6079c at https://pr26082-1c6079c.ngbuilds.io/.

mary-poppins commented:
You can preview 1742f11 at https://pr26082-1742f11.ngbuilds.io/.

petebacondarwin committed:
fix(ivy): ensure that `ɵNgModuleDefWithMeta` is exported

This is needed to build an app with `ngtsc` once `ngcc` has compiled
`@angular/core`.

mary-poppins commented:
You can preview e662200 at https://pr26082-e662200.ngbuilds.io/.

mary-poppins commented:
You can preview 15be1aa at https://pr26082-15be1aa.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/ngcc_host.ts:
```diff
@@ -37,4 +38,12 @@ export interface NgccReflectionHost extends ReflectionHost {
+   * @returns A collection of files objects that hold info about the decorated classes and import
+   * information.
+   */
+  findDecoratedFiles(entryPoint: ts.SourceFile): DecoratedFile[];
```
@alxhub I know that you will probably say that this is not really a `ReflectionHost` thing but it makes things much simpler to put it here - and let's just say it is a `NgccReflectionHost` thing :-)

mary-poppins commented:
You can preview 057c89d at https://pr26082-057c89d.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -294,6 +296,48 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+   * @returns A collection of files objects that hold info about the decorated classes and import
+   * information.
+   */
+  findDecoratedFiles(entryPoint: ts.SourceFile): DecoratedFile[] {
```
I don't mind the code as written for what it does.

What I'm not sure I understand is why `ngcc` is particularly focused on processing only those symbols which are reachable from the entry point.

Wouldn't it be 100x easier to do what `ngtsc` does, and iterate over all the files in the program via `ts.Program.getSourceFiles()`?

ngcc will need to not limit itself to symbols available from the entry point anyway to fix FW-519 (adding exports for modules/directives not exported from the entry point).

JoostK replied:
@alxhub In #26281, I believe @petebacondarwin started to switch to source file processing instead of only analyzing the public API surface. I don't have access to FW-519 I'm afraid, but it looks like #26281 addresses that issue (it's at least related).

petebacondarwin replied:
I agree that this is not the right approach, and as @JoostK says, there is already a PR that will change this approach. But I didn't want to start updating this PR when it was so close to being merged. Rest assured this API will not survive too long.

alxhub commented on packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -30,6 +30,9 @@ export interface DecorationAnalysis {
   constantPool: ConstantPool;
 }
 
+export type DecorationAnalyses = Map<ts.SourceFile, DecorationAnalysis>;
```
Clever!

I was going to suggest `class DeclarationAnalyses extends Map<...>` but apparently that doesn't work in ES5 :(

petebacondarwin replied:
I tried a number of approaches, and this appears to be the only one that actually works.

alxhub commented:
Thanks @JoostK for the thorough review!

alxhub commented:
http://test/OCL:216587597:BASE:216591069:1539206106033:b273a65b

jasonaden reviewed (approved):
Approve core change, adding private symbol to `r3_symbols.ts`

petebacondarwin commented:
Caretaker - the only outstanding CI failure is the flaky `aio_local` job.
